### ISSUE TITLE
Saving file from URL not working correctly

### ISSUE THREAD

**Issue by flyingL123** [czim/laravel-paperclip#33]
I'm trying to follow the instructions in the README for attaching a file from a URL, but it's not working correctly. It says:

> Stapler would automatically resolve any string URL assigned to the Model's attachment attribute and download and store the response to that URL. Paperclip does not do this automatically.
> If you wish to store the contents of a URL, you have some options. You can use the Czim\FileHandling\Storage\File\StorableFileFactory@makeFromUrl method and its return value.

My code looks like this:

```
$file = app(Czim\FileHandling\Storage\File\StorableFileFactory::class)->makeFromUrl($url);
$document = new App\ProductDocument;
$document->document = $file;
$document->save();
```

I'm getting the following error:

```
UnexpectedValueException with message 'Could not interpret given data, string value expected'
```

It's being thrown from Line 91 in `vendor/czim/file-handling/src/Storage/File/StorableFileFactory.php.`

Am I doing something wrong or is there a bug? I have confirmed that the file is being correctly downloaded. `$file` is an instance of `Czim\FileHandling\Storage\File\SplFileInfoStorableFile`. However, something goes wrong when trying to save the `$document`.

**Comment by czim** [comment]
Can you provide the content from your `filesystems.php` and `paperclip.php` configuration files? I have seen some cases where incomplete configuration caused this error.

**Comment by flyingL123** [comment]
Sure, thank you.


filesystems.php:

```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default Filesystem Disk
    |--------------------------------------------------------------------------
    |
    | Here you may specify the default filesystem disk that should be used
    | by the framework. The "local" disk, as well as a variety of cloud
    | based disks are available to your application. Just store away!
    |
    */

    'default' => env('FILESYSTEM_DRIVER', 'local'),

    /*
    |--------------------------------------------------------------------------
    | Default Cloud Filesystem Disk
    |--------------------------------------------------------------------------
    |
    | Many applications store files both locally and in the cloud. For this
    | reason, you may specify a default "cloud" driver here. This driver
    | will be bound as the Cloud disk implementation in the container.
    |
    */

    'cloud' => env('FILESYSTEM_CLOUD', 's3'),

    /*
    |--------------------------------------------------------------------------
    | Filesystem Disks
    |--------------------------------------------------------------------------
    |
    | Here you may configure as many filesystem "disks" as you wish, and you
    | may even configure multiple disks of the same driver. Defaults have
    | been setup for each driver as an example of the required options.
    |
    | Supported Drivers: "local", "ftp", "sftp", "s3", "rackspace"
    |
    */

    'disks' => [

        'local' => [
            'driver' => 'local',
            'root' => storage_path('app'),
        ],

        'public' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'url' => env('APP_URL').'/storage',
            'visibility' => 'public',
        ],

        's3' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
        ],

        'paperclip' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'visibility' => 'public',
        ],

        'paperclip-s3' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
            'visibility' => 'public',
        ]
    ],

];
```

And `paperclip.php`:

```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Model
    |--------------------------------------------------------------------------
    |
    | Configure how attached file information in stored on attachables.
    |
    */

    'model' => [

        // Mark which columns should be filled on the model by default.
        // These attributes are prefixed by <attribute name>_.
        'attributes' => [
            'size'         => true,
            'content_type' => true,
            'updated_at'   => true,
            'created_at'   => false,
            // JSON information about stored variants.
            'variants'     => false,
        ],

    ],

    /*
    |--------------------------------------------------------------------------
    | Storage
    |--------------------------------------------------------------------------
    |
    | Settings for handling storage of uploaded files.
    |
    */

    'storage' => [

        // The Laravel storage disk to use.
        'disk' => env('PAPERCLIP_DISK', 'paperclip'),

        // Per disk, the base URL where attachments are stored at
        'base-urls' => [
            'paperclip' => config('app.url') . '/storage',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Path
    |--------------------------------------------------------------------------
    |
    | The storage path that uploaded files and variants for models are placed in.
    |
    */

    'path' => [

        // The class that generates the paths
        'interpolator' => \Czim\Paperclip\Path\Interpolator::class,

        // The path to the original file to be interpolated. This will also\
        // be used for variant paths if the variant key is unset.
        'original' => ':class/:id_partition/:attribute/:variant/:filename',

        // If the structure for variant filenames should differ from the
        // original, it may be defined here.
        'variant'  => null,
    ],

    /*
    |--------------------------------------------------------------------------
    | Variants
    |--------------------------------------------------------------------------
    |
    | Processed files may have any number of variants: versions of the file that
    | are resized, rotated, compressed, or whatever you can think of.
    |
    */

    // The default to use for the main URL
    'default-variant' => 'original',

    // Variant processing configuration
    'variants' => [

        'aliases' => [
            'auto-orient' => \Czim\FileHandling\Variant\Strategies\ImageAutoOrientStrategy::class,
            'resize'      => \Czim\FileHandling\Variant\Strategies\ImageResizeStrategy::class,
        ],

        // If no specific variants are set for a clipped file on a Model, these
        // variant definitions will be used.
        'default' => [

            // 'variantname' => [
            //     'strategy-alias' => [ 'strategy' => 'configuration' ],
            // ],
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | File Preservation
    |--------------------------------------------------------------------------
    |
    | Settings to affect when attachment files are destroyed.
    |
    */

    // Set this to true in order to prevent older file uploads from being deleted.
    'keep-old-files' => false,

    // Set this to true in order to prevent file uploads from being deleted as attachments are destroyed.
    'preserve-files' => false,

    /*
    |--------------------------------------------------------------------------
    | Imagine
    |--------------------------------------------------------------------------
    |
    | The default binding to use for the ImagineInterface. May be Gd or Imagick.
    |
    */

    'imagine' => Imagine\Gd\Imagine::class,

    /*
    |--------------------------------------------------------------------------
    | Processing
    |--------------------------------------------------------------------------
    |
    | Settings for (re)processing attachments.
    |
    */

    'processing' => [
        'chunk-size' => 500,
    ],

];
```

**Comment by czim** [comment]
It's not any of the configuration, those files look totally fine.

I think I see what's going wrong now. And I think this is what's happening:

1. You use the StorableFileFactory to make an instance of a StorableFile. This is (admittedly confusingly) a data type that is part (only) of the internals of Paperclip.
2. Then, you pass this StorableFile instance into the attachment attribute of the model.
3. The model, on being `save()`d, passes the StorableFile instance back into the StorableFileFactory, which does not know what to make of it, since it is not an `SplFileInfo` or an `UploadedFile` instance, and finally falls back on assuming that a string is passed in with raw content for the file.

So, because you are doing the work Paperclip expects to have to do itself (create a StorableFile instance), it throws this error.

Long story short, you can probably fix this (for now) by passing the URL as a string into the attachment property directly, despite what the documentation says:

```php
$document = new App\ProductDocument;
$document->document = $url;
$document->save();
```
This should work for any case where`filter_var($url, FILTER_VALIDATE_URL)`is true.


This is clearly a fault in my documentation, that I will address as soon as I find the time to do so.
Additionally, I believe that passing any `StorableFile`, as you are doing, into the attachment property, should simply work! It's pretty silly that it does not, so I will make that work too. My bad.

**Comment by czim** [comment]
I've updated [czim/file-handling](https://github.com/czim/file-handling) (to version `1.0.4`).
This allows passing StorableFile instances into the attachment attribute.

I've also updated the paperclip documentation to remove the false claim that passing in a URL string will not work.
